### PR TITLE
Use caches for module resolution and type reference directives when using compiler default functions

### PR DESCRIPTION
**Pull request by sheetalkamat**
Uses #microsoft/TypeScript#43700

### CODE REVIEW

**Comment by JasonKleban** [comment]
@johnnyreilly could this please be also get applied to version 8.x for webpack 4 compatibility?

**Comment by JasonKleban** [comment]
What is the technique for testing both this and the typescript branch locally at once?

**Comment by johnnyreilly** [comment]
> @johnnyreilly could this please be also get applied to version 8.x for webpack 4 compatibility?

If you wanted to do a PR to backport it against the https://github.com/TypeStrong/ts-loader/tree/webpack-4 branch after this has been merged then we could potentially look at it.

**Comment by sheetalkamat** [comment]
@johnnyreilly I am not sure what the test failures mean but i am able to successfully run the tests locally.
This is ready to merge pending review.

**Comment by sheetalkamat** [comment]
@JasonKleban I had to manually create links for typescript built locally while I was working on this and ts-loader as well and use it in your sample repro.  That's how I tested it, not sure if that's how you would need to test but bot should post typescript build from the PR at https://github.com/microsoft/TypeScript/pull/43700#issuecomment-824347567

**Comment by johnnyreilly** [comment]
> @johnnyreilly I am not sure what the test failures mean but i am able to successfully run the tests locally.
> This is ready to merge pending review.

Thanks @sheetalkamat - looking at the test failure it looked like a random yarn failure. I've requeued and will try to make time to do the review soon. Thanks for your efforts!

**Comment by JasonKleban** [comment]
@sheetalkamat can you share any indication for how big of a performance impact you expect this to have?  Might it nearly equalize ts-loader & webpack to tsc compilation times?

**Comment by sheetalkamat** [comment]
> Might it nearly equalize ts-loader & webpack to tsc compilation times?

@JasonKleban probably.. The sample was such a small, that i couldnt measure it reliabily especially with my local setup but i did verify that lot of file accesses and checks get reduced (esp you can see that in trace logs). Because of the way ts-loader works(i am not sure if it is intentional or not but i didnt go about investigating that) i have seem that program is created with some subset of files and more files get added to it as webpack asks about more information about dependencies..  So in that second pass, you start reusing module resolution, type reference resolution, package json cache etc... Not that packagejson improvement can be seen (from trace logs) for semver types as well as it tries to read packagejson many times the first time itself.

File: src/interfaces.ts
Comment (johnnyreilly) [inline src/interfaces.ts:172]: So this will be accessible in the TypeScript API in future?

File: src/servicesHost.ts
Comment (johnnyreilly) [inline src/servicesHost.ts:1173]:
```suggestion
// Until the api is published
```

File: src/servicesHost.ts
Comment (johnnyreilly) [inline src/servicesHost.ts:1186]:
```suggestion
      // Until the api is published
```

File: src/interfaces.ts
Comment (johnnyreilly) [reply src/interfaces.ts:172]: I think this may be it: https://github.com/microsoft/TypeScript/blob/74993a2a64bb2e423b40204bb54ff749cdd4ef54/src/compiler/moduleNameResolver.ts#L458

Comment (johnnyreilly) [reply src/interfaces.ts:172]:
```suggestion
// Until the API has been released and ts-loader is built against a version of TypeScript that contains it - see https://github.com/microsoft/TypeScript/blob/74993a2a64bb2e423b40204bb54ff749cdd4ef54/src/compiler/moduleNameResolver.ts#L458
```

**Review by johnnyreilly — APPROVED** [review]
This looks great!  Fine work @sheetalkamat!

I'm happy for this to be released and I've updated the `package.json` / `CHANGELOG.md` accordingly.

File: test/comparison-tests/projectReferencesSymLinks/expectedOutput-4.1/patch0/output.txt
Comment (sokra) [inline test/comparison-tests/projectReferencesSymLinks/expectedOutput-4.1/patch0/output.txt:9]: hmm? why is this a backslash?

btw. webpack usually would display this as `ERROR in ./app/src/index.ts`

Comment (johnnyreilly) [reply test/comparison-tests/projectReferencesSymLinks/expectedOutput-4.1/patch0/output.txt:9]: I suspect this is just ts-loader surfacing up what TypeScript hands over:

https://github.com/TypeStrong/ts-loader/blob/c2d1ca2d5ec6fbca3f4b8c5e1e755e61af719a84/src/utils.ts#L77

So if you're running on a Windows machine it's likely `\` country

**Comment by felixmosh** [comment]
I'm getting an error with v9.1.0-9.1.1, since this is the only change in these versions, I post it here

```shell
[webpack-cli] HookWebpackError: Cannot read property 'length' of undefined
    at makeWebpackError (/Users/***/Projects/node/***/node_modules/webpack/lib/HookWebpackError.js:49:9)
    at /Users/***/Projects/node/***/node_modules/webpack/lib/Compilation.js:2496:12
    at eval (eval at create (/Users/***/Projects/node/***/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:26:1)
    at fn (/Users/***/Projects/node/***/node_modules/webpack/lib/Compilation.js:427:17)
    at Hook.eval [as callAsync] (eval at create (/Users/***/Projects/node/***/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:24:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/Users/***/Projects/node/***/node_modules/tapable/lib/Hook.js:18:14)
    at cont (/Users/***/Projects/node/***/node_modules/webpack/lib/Compilation.js:2493:34)
    at /Users/***/Projects/node/***/node_modules/webpack/lib/Compilation.js:2539:10
    at /Users/***/Projects/node/***/node_modules/neo-async/async.js:2830:7
    at Object.each (/Users/***/Projects/node/***/node_modules/neo-async/async.js:2850:39)
-- inner error --
TypeError: Cannot read property 'length' of undefined
    at computeLineStarts (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:9580:27)
    at Object.getLineStarts (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:9640:60)
    at getCurrentLineMap (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:101516:59)
    at emitDetachedCommentsAndUpdateCommentsInfo (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:105227:94)
    at emitBodyWithDetachedComments (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:105067:17)
    at emitSourceFile (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:103774:21)
    at pipelineEmitWithHint (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:101597:24)
    at noEmitNotification (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:100109:9)
    at onEmitNode (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:85991:13)
    at onEmitNode (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:96421:17)
caused by plugins in Compilation.hooks.processAssets
TypeError: Cannot read property 'length' of undefined
    at computeLineStarts (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:9580:27)
    at Object.getLineStarts (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:9640:60)
    at getCurrentLineMap (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:101516:59)
    at emitDetachedCommentsAndUpdateCommentsInfo (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:105227:94)
    at emitBodyWithDetachedComments (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:105067:17)
    at emitSourceFile (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:103774:21)
    at pipelineEmitWithHint (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:101597:24)
    at noEmitNotification (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:100109:9)
    at onEmitNode (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:85991:13)
    at onEmitNode (/Users/***/Projects/node/***/node_modules/typescript/lib/typescript.js:96421:17)
error Command failed with exit code 2.
```

**Pull request merged**
